packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -564,16 +591,59 @@ export class ApolloServerBase {
         )
       ).filter(
         (maybeServerListener): maybeServerListener is GraphQLServerListener =>
-          typeof maybeServerListener === 'object' &&
-          !!maybeServerListener.serverWillStop,
+          typeof maybeServerListener === 'object',
       );
-      this.toDispose.add(async () => {
-        await Promise.all(
-          serverListeners.map(({ serverWillStop }) => serverWillStop?.()),
-        );
+
+      serverListeners.forEach(({ schemaDidLoadOrUpdate, serverWillStop }) => {
+        if (schemaDidLoadOrUpdate) {
+          if (maybeGateway) {
+            if (!maybeGateway.onSchemaLoadOrUpdate) {
+              // TODO: Once a gateway version providing the core schema to
+              //       callbacks has been released, update this message to state
+              //       the specific version needed.
+              throw new Error(
+                [
+                  `One of your plugins uses the 'onSchemaLoadOrUpdate' hook,`,
+                  `but your gateway version is too old to support this hook.`,
+                  `Please update your gateway version to latest.`,
+                ].join(' '),
+              );
+            }
+            // This check shouldn't ever fail, but Typescript assumes this.state
+            // can be anything at this point, so we have this to help Typescript.
+            if (this.state.phase !== 'invoking serverWillStart') {
+              throw new Error(`State was unexpectedly ${this.state.phase}`);
+            }
+            // It is crucial there not be any await statements between this
+            // read and the addition of a listener to the listener set, or else
+            // schema updates could be missed by listeners.
+            const latestSchemaDerivedData = this.state.schemaDerivedData;
```

Comment:
This is all far too complicated for me to understand or have any hope of successfully maintaining. Are there other approaches we can use to deal with the possibility of receiving schema change updates during startup?

For example, could we just buffer them?  If a schema change comes in during 'invoking serverWillStart' we save it to some field (and I think it's fine if we just have one field rather than an array — if two schema changes come in we can just ignore the non-last one), and after serverWillStart is done we can fire off any buffered schema?

ie, that would let us just consistently use `const schemaDerivedData` for all relevant calls during the course of `_start`. And instead of changing the hook from "ignore in non-started state" to "ignore in all but three states", it would be "buffer in the starting/serverWillStart states, call in started, ignore otherwise".

Reply:
>This is all far too complicated for me to understand or have any hope of successfully maintaining. Are there other approaches we can use to deal with the possibility of receiving schema change updates during startup?

In the refactor, I've tried to address this by confining most of this code to non-async functions, which I feel makes it easier to reason about since such functions can't be suspended during execution.

>For example, could we just buffer them? If a schema change comes in during 'invoking serverWillStart' we save it to some field (and I think it's fine if we just have one field rather than an array — if two schema changes come in we can just ignore the non-last one), and after serverWillStart is done we can fire off any buffered schema?

I added some buffering in the `SchemaManager` class, there's more details about it in the `SchemaManager.onSchemaLoadOrUpdate` docstring.